gc/service/service.go

```diff
@@ -49,91 +51,153 @@ func (s *service) Run(ctx context.Context) error {
 		now := s.cfg.TimeNowFunc().UTC()
 
 		for _, container := range containers {
+			if container.VersionsTTL < 0 {
+
+				err = s.deleteExpiredUnpublishedVersions(ctx, now, namespace, container)
+				if err != nil {
+					return errors.Wrapf(err, "error deleting expired unpublished versions for container `%s/%s`", namespace, container.Name)
+				}
+			} else {
+				err = s.deleteExpiredVersions(ctx, now, namespace, container)
+				if err != nil {
+					return errors.Wrapf(err, "error deleting expired versions for container `%s/%s`", namespace, container.Name)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (s *service) deleteExpiredVersions(ctx context.Context, now time.Time, namespace string, container models.Container) error {
+	log.WithFields(log.Fields{
+		"namespace": namespace,
+		"container": container,
+	}).Debugf("listing expired versions ...")
+
+	versions, err := s.cfg.MdRepo.ListAllVersionsByContainer(ctx, namespace, container.Name)
+	if err != nil {
+		return errors.Wrapf(err, "error listing versions for container `%s/%s`", namespace, container)
+	}
+
+	for _, version := range versions {
+		log.WithFields(log.Fields{
+			"namespace": namespace,
+			"container": container,
+			"version":   version.Name,
+		}).Debugf("listing expired versions ...")
+
+		if version.CreatedAt.After(now.Add(-1 * container.VersionsTTL)) {
```

Comment:
I would prefer to use positive filters i.e. if true ; do delete ... ; done so to log things to be deleted instead ones not going to be deleted.

This logic looks more straight forward